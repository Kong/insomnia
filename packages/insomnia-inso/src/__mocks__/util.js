// eslint-disable-next-line filenames/match-exported
const mod = jest.requireActual('../util');

mod.exit = jest.fn();

module.exports = mod;
