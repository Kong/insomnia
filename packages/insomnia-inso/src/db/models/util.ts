import { InsoError } from '../../errors';
import type { BaseModel } from './types';

export const matchIdIsh = ({ _id }: BaseModel, identifier: string) =>
  _id.startsWith(identifier);

export const generateIdIsh = ({ _id }: BaseModel, length = 10) =>
  _id.substr(0, length);

function indent(level: number, code: string, tab = '  |'): string {
  if (!level || level < 0) {
    return code;
  }

  const prefix = new Array(level + 1).join(tab);
  return `${prefix} ${code}`;
}

export const getDbChoice = (
  idIsh: string,
  message: string,
  config: {
        indent?: number
        hint?: string
    } = {},
) => ({
  name: idIsh,
  message: indent(config.indent || 0, message),
  value: `${message} - ${idIsh}`,
  hint: config.hint || `${idIsh}`,
});

export const ensureSingleOrNone = <T>(
  items: T[],
  entity: string,
): T | null | undefined => {
  if (items.length === 1) {
    return items[0];
  }

  if (items.length === 0) {
    return null;
  }

  throw new InsoError(
    `Expected single or no ${entity} in the data store, but found multiple (${items.length}).`,
  );
};
export const ensureSingle = <T>(items: T[], entity: string): T => {
  if (items.length === 1) {
    return items[0];
  }

  if (items.length === 0) {
    throw new InsoError(
      `Expected single ${entity} in the data store, but found none.`,
    );
  }

  throw new InsoError(
    `Expected single ${entity} in the data store, but found multiple (${items.length}).`,
  );
};
