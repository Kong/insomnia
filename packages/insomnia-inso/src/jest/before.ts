import { logType } from 'consola';

import { logger, LogsByType } from '../logger';

// Taken from https://github.com/unjs/consola/blob/master/types/consola.d.ts#L16
const logTypes: logType[] = [
  'silent',
  'fatal',
  'error',
  'warn',
  'log',
  'info',
  'success',
  'debug',
  'trace',
  'verbose',
  'ready',
  'start',
];

export const globalBeforeAll = () => {
  logger.__getLogs = () => logTypes.reduce((accumulator, level) => {
    // @ts-expect-error the consola types are incomplete
    const calls = (logger[level] as jest.Mock).mock.calls.map(call => (
      call.length === 1 ? call[0] : call
    ));
    return {
      ...accumulator,
      [level]: calls,
    };
  }, {} as LogsByType);
};

export const globalBeforeEach = () => {
  logger.mockTypes(() => jest.fn());
};
