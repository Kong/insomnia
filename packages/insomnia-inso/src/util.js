// @flow
import * as packageJson from '../package.json';
import { handleError } from './errors';

export function getVersion() {
  return isDevelopment() ? 'dev' : packageJson.version;
}

export function isDevelopment() {
  return process.env.NODE_ENV === 'development';
}

export function logErrorExit1(err?: Error) {
  handleError(err);
  process.exit(1);
}

export async function exit(result: Promise<boolean>): Promise<void> {
  return result.then(r => process.exit(r ? 0 : 1)).catch(logErrorExit1);
}

export function getDefaultAppDataDir(): string {
  const dir = process.env.DEFAULT_APP_DATA_DIR;

  if (!dir) {
    throw new Error('Environment variable DEFAULT_APP_DATA_DIR is not set.');
  }

  return dir;
}
