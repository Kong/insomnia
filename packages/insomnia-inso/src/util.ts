import packageJson from '../package.json';
import { handleError } from './errors';

export const getVersion = () => {
  return isDevelopment() ? 'dev' : packageJson.version;
};

export const isDevelopment = () => {
  return process.env.NODE_ENV === 'development';
};

export const logErrorExit1 = (err?: Error) => {
  handleError(err);
  process.exit(1);
};

export const exit = async (result: Promise<boolean>): Promise<void> => {
  return result.then(r => process.exit(r ? 0 : 1)).catch(logErrorExit1);
};

export const getDefaultAppName = (): string => {
  const name = process.env.DEFAULT_APP_NAME;

  if (!name) {
    throw new Error('Environment variable DEFAULT_APP_NAME is not set.');
  }

  return name;
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const noop = () => {};
