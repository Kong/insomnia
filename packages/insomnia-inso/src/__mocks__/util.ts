import { jest } from '@jest/globals';

import * as utilOriginal from '../util';

// eslint-disable-next-line filenames/match-exported
const util = jest.requireActual('../util') as typeof utilOriginal;
util.exit = jest.fn();

module.exports = util;
