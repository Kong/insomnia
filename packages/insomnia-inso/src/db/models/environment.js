// @flow
import type { Database } from '../index';
import type { Environment } from './types';
import { mustFindSingle } from '../index';
import { AutoComplete } from 'enquirer';
import { generateIdIsh, getDbChoice, matchIdIsh } from './util';

export const loadEnvironment = (
  db: Database,
  workspaceId: string,
  identifier?: string,
): ?Environment => {
  if (!db.Environment.length) {
    return null;
  }

  // Get the sub environments
  const baseWorkspaceEnv = mustFindSingle(db.Environment, e => e.parentId === workspaceId);
  const subEnvs = db.Environment.filter(e => e.parentId === baseWorkspaceEnv._id);

  // try to find a sub env, otherwise return the base env
  return identifier && subEnvs.length
    ? subEnvs.find(e => matchIdIsh(e, identifier) || e.name === identifier)
    : baseWorkspaceEnv;
};

export const promptEnvironment = async (
  db: Database,
  ci: boolean,
  workspaceId: string,
): Promise<?Environment> => {
  if (ci || !db.Environment.length) {
    return null;
  }

  // Get the sub environments
  const baseWorkspaceEnv = mustFindSingle(db.Environment, e => e.parentId === workspaceId);
  const subEnvs = db.Environment.filter(e => e.parentId === baseWorkspaceEnv._id);

  if (!subEnvs.length) {
    return baseWorkspaceEnv;
  }

  const prompt = new AutoComplete({
    name: 'environment',
    message: `Select an environment`,
    choices: subEnvs.map(e => getDbChoice(generateIdIsh(e, 14), e.name)),
  });

  const [idIsh] = (await prompt.run()).split(' - ').reverse();
  return mustFindSingle(db.Environment, e => matchIdIsh(e, idIsh));
};
