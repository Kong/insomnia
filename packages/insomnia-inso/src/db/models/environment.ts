// @ts-expect-error the enquirer types are incomplete https://github.com/enquirer/enquirer/pull/307
import { AutoComplete } from 'enquirer';

import { logger } from '../../logger';
import type { Database } from '../index';
import type { Environment } from './types';
import { ensureSingle, generateIdIsh, getDbChoice, matchIdIsh } from './util';

const loadBaseEnvironmentForWorkspace = (db: Database, workspaceId: string): Environment => {
  logger.trace(
    'Load base environment for the workspace `%s` from data store',
    workspaceId,
  );
  const items = db.Environment.filter(e => e.parentId === workspaceId);
  logger.trace('Found %d.', items.length);
  return ensureSingle(items, 'base environment');
};

export const loadEnvironment = (
  db: Database,
  workspaceId: string,
  identifier?: string,
): Environment | null | undefined => {
  if (!db.Environment.length) {
    return null;
  }

  // Get the sub environments
  const baseWorkspaceEnv = loadBaseEnvironmentForWorkspace(db, workspaceId);
  const subEnvs = db.Environment.filter(
    env => env.parentId === baseWorkspaceEnv._id,
  );

  // If no identifier, return base environment
  if (!identifier) {
    logger.trace('No sub environments found, using base environment');
    return baseWorkspaceEnv;
  }

  logger.trace(
    'Load sub environment with identifier `%s` from data store',
    identifier,
  );
  const items = subEnvs.filter(
    env => matchIdIsh(env, identifier) || env.name === identifier,
  );
  logger.trace('Found %d', items.length);
  return ensureSingle(items, 'sub environment');
};

export const promptEnvironment = async (
  db: Database,
  ci: boolean,
  workspaceId: string,
): Promise<Environment | null | undefined> => {
  if (ci || !db.Environment.length) {
    return null;
  }

  // Get the sub environments
  const baseWorkspaceEnv = loadBaseEnvironmentForWorkspace(db, workspaceId);
  const subEnvs = db.Environment.filter(
    e => e.parentId === baseWorkspaceEnv._id,
  );

  if (!subEnvs.length) {
    logger.trace('No sub environments found, using base environment');
    return baseWorkspaceEnv;
  }

  const prompt = new AutoComplete({
    name: 'environment',
    message: 'Select an environment',
    choices: subEnvs.map(e => getDbChoice(generateIdIsh(e, 14), e.name)),
  });
  logger.trace('Prompt for environment');
  const [idIsh] = (await prompt.run()).split(' - ').reverse();
  return loadEnvironment(db, workspaceId, idIsh);
};
