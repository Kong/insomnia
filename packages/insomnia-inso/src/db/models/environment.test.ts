import _enquirer from 'enquirer';

import { globalBeforeAll, globalBeforeEach } from '../../jest/before';
import type { Database } from '../index';
import { emptyDb } from '../index';
import { loadEnvironment, promptEnvironment } from './environment';
import type { Environment, Workspace } from './types';
import { generateIdIsh } from './util';

jest.mock('enquirer');
const enquirer = _enquirer as jest.MockedClass<typeof _enquirer> & {
  __mockPromptRun: (str: string) => void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- missing types from enquirer
  __constructorMock: jest.Mock;
};

describe('Environment', () => {
  beforeAll(() => {
    globalBeforeAll();
  });

  beforeEach(() => {
    globalBeforeEach();
  });

  let db: Database = emptyDb();

  const workspace = {
    _id: 'wrk_1234567890',
    name: 'workspace name',
  } as Workspace;

  const environment = {
    _id: 'env_1234567890',
    name: 'base env',
    parentId: workspace._id,
  } as Environment;

  const subEnvironment = {
    _id: 'env_env_1234567890',
    name: 'sub env',
    parentId: environment._id,
  } as Environment;

  beforeEach(() => {
    db = emptyDb();

    const dummySubEnv = {
      _id: 'env_env_dummy',
      name: 'dummy sub env',
      parentId: environment._id,
    } as Environment;

    db.Workspace.push(workspace);
    db.Environment.push(environment);
    db.Environment.push(subEnvironment);
    db.Environment.push(dummySubEnv);
    jest.clearAllMocks();
  });

  describe('promptEnvironment()', () => {
    it('should return null if ci', () => {
      expect(promptEnvironment(db, true, workspace._id)).resolves.toBeNull();
    });

    it('should return null if no environments exist', () => {
      db.Environment = [];
      expect(promptEnvironment(db, false, workspace._id)).resolves.toBeNull();
    });

    it('should throw error if base env for workspace not found', () => {
      expect(promptEnvironment(db, false, 'workspace-not-found')).rejects.toThrowError();
    });

    it('should throw error if multiple base env for workspace found', () => {
      db.Environment.push({ ...environment });
      expect(promptEnvironment(db, false, workspace._id)).rejects.toThrowError();
    });

    it('should load sub environment after prompt result', async () => {
      enquirer.__mockPromptRun('environment - env_env_123456');

      expect(promptEnvironment(db, false, workspace._id)).resolves.toBe(subEnvironment);
    });

    it('should match snapshot of autocomplete config', async () => {
      await promptEnvironment(db, false, workspace._id);
      expect(enquirer.__constructorMock.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('loadEnvironment()', () => {
    it('should return null if no environments exist', () => {
      db.Environment = [];
      expect(loadEnvironment(db, workspace._id, environment._id)).toBeNull();
    });

    it('should throw error if base env for workspace not found', () => {
      expect(() => loadEnvironment(db, 'workspace-not-found', environment._id)).toThrowError();
    });

    it('should throw error if multiple base env for workspace found', () => {
      db.Environment.push({ ...environment });
      expect(() => loadEnvironment(db, workspace._id, environment._id)).toThrowError();
    });

    it.each([generateIdIsh(subEnvironment), subEnvironment._id, subEnvironment.name])(
      'should return the sub environment if matched with id: %s',
      () => {
        expect(loadEnvironment(db, workspace._id, subEnvironment._id)).toBe(subEnvironment);
      },
    );

    it('should return the base environment if env id not specified', () => {
      expect(loadEnvironment(db, workspace._id)).toBe(environment);
    });

    it('should return the base environment if no sub envs exist', () => {
      db.Environment = [environment];
      expect(loadEnvironment(db, workspace._id)).toBe(environment);
    });

    it('should throw error sub env not found', () => {
      db.Environment = [environment];
      expect(() => loadEnvironment(db, workspace._id, subEnvironment._id)).toThrowError();
    });
  });
});
