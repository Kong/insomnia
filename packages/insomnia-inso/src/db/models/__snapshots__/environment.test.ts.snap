// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Environment promptEnvironment() should match snapshot of autocomplete config 1`] = `
Object {
  "choices": Array [
    Object {
      "hint": "env_env_123456",
      "message": "sub env",
      "name": "env_env_123456",
      "value": "sub env - env_env_123456",
    },
    Object {
      "hint": "env_env_dummy",
      "message": "dummy sub env",
      "name": "env_env_dummy",
      "value": "dummy sub env - env_env_dummy",
    },
  ],
  "message": "Select an environment",
  "name": "environment",
}
`;
